src/core/lib/security/credentials/oauth2/oauth2_credentials.cc

```diff
@@ -472,6 +474,140 @@ grpc_call_credentials* grpc_google_refresh_token_credentials_create(
+
+ protected:
+  void fetch_oauth2(grpc_credentials_metadata_request* metadata_req,
+                    grpc_httpcli_context* http_context,
+                    grpc_polling_entity* pollent,
+                    grpc_iomgr_cb_func response_cb,
+                    grpc_millis deadline) override {
+    grpc_http_header header = {(char*)"Content-Type",
+                               (char*)"application/x-www-form-urlencoded"};
+    grpc_httpcli_request request;
+    char* body = nullptr;
+    gpr_asprintf(&body, GRPC_STS_POST_MINIMAL_BODY_FORMAT_STRING,
+                 subject_token_.get(), subject_token_type_.get());
+    body = maybe_add_to_query(body, "resource", resource_.get());
+    body = maybe_add_to_query(body, "audience", audience_.get());
+    body = maybe_add_to_query(body, "scope", scope_.get());
+    body = maybe_add_to_query(body, "requested_token_type",
+                              requested_token_type_.get());
+    body = maybe_add_to_query(body, "actor_token", actor_token_.get());
+    body =
+        maybe_add_to_query(body, "actor_token_type", actor_token_type_.get());
+
+    memset(&request, 0, sizeof(grpc_httpcli_request));
+    request.host = (char*)sts_url_->authority;
+    request.http.path = (char*)sts_url_->path;
+    request.http.hdr_count = 1;
+    request.http.hdrs = &header;
+    request.handshaker = (strcmp(sts_url_->scheme, "https") == 0)
+                             ? &grpc_httpcli_ssl
+                             : &grpc_httpcli_plaintext;
+    /* TODO(ctiller): Carry the resource_quota in ctx and share it with the host
+       channel. This would allow us to cancel an authentication query when under
+       extreme memory pressure. */
+    grpc_resource_quota* resource_quota =
+        grpc_resource_quota_create("oauth2_credentials_refresh");
+    grpc_httpcli_post(http_context, pollent, resource_quota, &request, body,
+                      strlen(body), deadline,
+                      GRPC_CLOSURE_CREATE(response_cb, metadata_req,
+                                          grpc_schedule_on_exec_ctx),
+                      &metadata_req->response);
+    grpc_resource_quota_unref_internal(resource_quota);
+    gpr_free(body);
+  }
+
+ private:
+  grpc_uri* sts_url_;
+  grpc_core::UniquePtr<char> resource_;
+  grpc_core::UniquePtr<char> audience_;
+  grpc_core::UniquePtr<char> scope_;
+  grpc_core::UniquePtr<char> requested_token_type_;
+  grpc_core::UniquePtr<char> subject_token_;
+  grpc_core::UniquePtr<char> subject_token_type_;
+  grpc_core::UniquePtr<char> actor_token_;
+  grpc_core::UniquePtr<char> actor_token_type_;
+};
+
+}  // namespace
+
+grpc_uri* grpc_validate_sts_credentials_options(
+    const grpc_sts_credentials_options* options) {
+  grpc_uri* sts_url = nullptr;
+  if (options->sts_endpoint_url == nullptr ||
+      strlen(options->sts_endpoint_url) == 0) {
+    gpr_log(GPR_ERROR, "sts_endpoint_url needs to be specified");
+    goto fail;
```

Comment:
nit: you can directly return nullptr here.

Reply:
I thought about it but decided against it for the sake of consistency in the error handling pattern in this function. I can still do it if you feel strongly about it.

Reply:
SG